### DISCUSSION TITLE
AOSP Indexing take long time

### DISCUSSION THREAD

**Discussion by JustSong** [oracle/opengrok#4038] [Q&A] [answered]
**Version:**
- Tomcat: apache-tomcat-8.5.37.tar.gz
- Opengrok: opengrok-1.3.10.tar.gz
- Ctags: Exuberant Ctags 5.8

**Index command:**
```
java -Djava.util.logging.config.file=$OPENGROK_LOG_FILE_PATH \
     -jar $OPENGROK_PROJECT_PATH/$OPENGROK_VERSION/lib/opengrok.jar \
     -s ${OPENGROK_SRC_ROOT} \
     -d ${OPENGROK_DATA_ROOT} \
     -c ${OPENGROK_CTAGS} \
     -W ${OPENGROK_XML_CONFIGURATION} \
     -U ${OPENGROK_WEBAPP_CFGADDR} \
     --depth 100 \
     -S -H -P -G --threads 8
```

When I index a large size source code (AOSP-Android10), it takes almost 24 hours to complete.
I have checked the log and find it will take so long time after every "Done optimizing index for project".

Example Log:
> 2022-09-12 17:16:53.439+0800 FINE t15307945 PendingFileCompleter.complete: renamed 42 file(s) (took 0 seconds)
2022-09-12 17:16:53.439+0800 FINE t15307945 PendingFileCompleter.complete: affirmed links for 0 path(s) (took 0 seconds)
2022-09-12 17:16:53.439+0800 FINE t15307945 IndexDatabase.finishWriting: completed 43 object(s)
2022-09-12 17:16:53.464+0800 INFO t15307945 IndexDatabase.optimize: Optimizing the index for project device
2022-09-12 **17:16:57**.568+0800 INFO t15307945 Statistics.report: **Done optimizing index for project** device (took 4.104 seconds)
2022-09-12 **18:46:21**.227+0800 FINE t15307944 HttpURLConnection.getInputStream0: sun.net.www.MessageHeader@748051322 pairs: {null: HTTP/1.1 204}{Date: Mon, 12 Sep 2022 10:46:21 GMT}

Question 1: The index time: 24 hours, is this normal index time?

Question 2: Is there any solution to speed up the AOSP indexing time?
EX: Do we need to set a special setting when we index AOSP source?

**Comment by vladak** [comment, marked as answer]
Thanks to the indexer being heavily parallelized, the indexing time depends very much on the underlying hardware and the resources assigned to the indexer's JVM.

Some pieces of advice:
  - use the latest OpenGrok version
  - increase thread count (assuming the machine has ample CPU power)
  - increase Java heap (assuming enough real RAM)

Also check out https://github.com/oracle/opengrok/wiki/Tuning-for-large-code-bases

**Reply by vladak** [reply]
As for the 1.5 hr delay between `IndexDatabase.optimize()` (that has been removed in the latest OpenGrok release) and the configuration being sent to the webapp, this is really strange. Could you share the complete end of the log ?

**Reply by JustSong** [reply]
Thanks a lot for your advice, I will try to use a better performance PC to index. If I have any improve I will that you know.

**Reply by JustSong** [reply]
- JDK: Java11.0.16.1
- Web server: Tomcat10.0.23
- Ctags: Universal Ctags 5.9.0(aaf5fa4f)
- Opengrok: 1.7.35

After I updated the opengrok version, we got a huge improve for indexing time.
The first time for indexing only takes 3hrs. Reindex with no-changes source only take 20-40min.

Thanks again for @vladak  advices

**Comment by chenchuanliang** [comment]
I have the same problem. We need to index more than 60 items.From android 7 to android 12.It still takes a long time without new projects

**Reply by JustSong** [reply]
Have you tried to upgrade OpenGrok to the latest version? It's really helped me a lot.

**Reply by yanggithuv** [reply]
with the opengrok opengrok-1.11.1 ，need 1 day with the first index，do you have ant ideas about this

**Reply by vladak** [reply]
The duration of the indexing, and initial indexing especially, depends on multiple factors. Firstly, it depends on how beefy the machine running the indexer is. Generally, the more CPUs and RAM and fast I/O you throw at the indexer, the faster it will get as it is heavily concurrent. Secondly, it depends on how the indexer is configured and how it is run. If the machine has ample RAM but the Java heap is small, the indexer might be progressing slowly (or eventually run of out heap space, crashing in not so obvious ways, or even hanging indefinitely - such is the life of Java programs sometimes). Then it depends on the indexer features enabled for given input data. Things like whether merge changesets or renamed files are enabled, if history cache is enabled, if the economy mode for xrefs is active, etc.

AOSP is definitely considered large code base, so https://github.com/oracle/opengrok/wiki/Tuning-for-large-code-bases should be used for advice.

So, if anyone has access to well spec'd machine or VM (say more than 128 CPUs and more than 512 GB of RAM with some nice NVME based I/O backend with crazy IOPS), I'd be curious how fast the initial AOSP index time can be (assuming indexer tuning is done). Let the races begin ! :-)

**Reply by vladak** [reply]
For the record, using a VM in Oracle OCI with the following specs:
  - OS: CentOS 8 Stream 2022.10.26-0 (updated)
    - paravirtualized
  - OCI shape: VM.Standard.E4.Flex
    - 64 CPUs (AMD - `/proc/cpuinfo` says AMD EPYC 7J13)
    - 512 GB RAM
    - boot volume 50 GB, UHP (VPU/GB:50), 50,000 IOPS, Throughput 480 MB/s
    - block volume 1024 GB, UHP (VPU/GB:50), 122,880 IOPS, 983.04 MB/s
      - for OpenGrok data

and no other special system tuning, running the indexer like this:
```
java -Xmx384G -jar /var/opengrok/dist/opengrok-1.12.3/lib/opengrok.jar\
    -c /usr/local/bin/ctags \
    -P -H -S -G --progress -s /var/opengrok/src -d /var/opengrok/data -W \
    /var/opengrok/etc/configuration.xml
```
using OpenGrok 1.12.3, it performed initial indexing of AOSP in 1:40:44 - that's 1 hour and 40 minutes.

Looking at the utilization in the Oracle OCI Metrics view, neither CPU or RAM were fully loaded, rather it the low percentages (20-40%), so there is probably space for tuning the indexer (increasing the number of threads, buffer size for Lucene, increase the number of commits to process in chunks for the per partes history cache generation, etc).

For reference, the data sizes were as follows after the indexing:
```
24G     historycache
26G     index
13G     xref
```
The overall size of the source root directory was 172G - AOSP was the only project present.

There was a significant progress slowdown at the very last phase of indexing (like the last one hour or so), the CPU usage was very low at that point. Basically 2 CPUs were pegged inside `org.opengrok.indexer.analysis.plain.XMLXref.yylex()` for very long, other CPUs were idle. That might be worth looking into as some input files probably caused the lexer to spin for a long time. Solving this would reduce the indexing time significantly. Could be related to #907 or #3740.

**Reply by vladak** [reply]
After adding xref timeout to XMLAnalyzer (PR #4328), the **initial** AOSP indexing time using the above setup got reduced to 35 minutes.

The fix (sic!) will be part of 1.12.4.

**Reply by vladak** [reply]
Building on top of the `XMLAnalyzer` changes, I tried to add `--threads 256 --historyThreads 256` to the indexer arguments. This had marginal effect on the indexing time.

**Reply by vladak** [reply]
Tried with `--threads 512 --historyThreads 512`, no change either. Also used this with bigger amount of heap (`-Xmx484G`), no improvement.

Looking at the CPU utilization near the end of indexing, even after the `XMLAnalyzer` changes, the CPU utilization is still low there, so perhaps there is more room for improvement.

**Reply by vladak** [reply]
Re-running the original indexer invocation with just released 1.12.4 and captures whole indexer run: 
![image](https://user-images.githubusercontent.com/2934284/235707826-ccad769c-0c63-427c-9899-6ff466c313ab.png)

There is one repository with larger history that leads to the long tail in the first phase of indexing (which took ~20 minutes, the end visible as the blue triangle annotation in the CPU pane) and there is also some trailing processing in the 2nd phase of the indexing, which might be worth further investigation.

**Reply by vladak** [reply]
I tried setting `historyChunkCount` tunable to 50000000 (the default for Git is to process 65536 changesets and repeat until the end of history for given repo) to process all commits from the trailing repository in one go, did not have much effect w.r.t. overall indexing time (0:32:47), maybe shaving couple of minutes however I'd have to re-run the indexer with such settings multiple times in order to be sure.

Here's the whole indexer run graph for comparison: 
![image](https://user-images.githubusercontent.com/2934284/235713082-292487ca-a895-42dd-9f34-6e838f1973cd.png)

**Comment by vladak** [comment]
Another improvement for AOSP indexing is being done in PR #4552. This is for a specific set of settings (renamed file handling on, annotation cache on) however might speed up general AOSP reindex time.
